File: algorithms/_test/estrutura-de-dados-arrays-strings.py

# Arrays e Strings:

# # Inverter string
s = "hello"
print(s[::-1])  # 'olleh'


# # Encontrar subarray com soma máxima (Kadane)
def max_subarray(nums):
    max_sum = curr = nums[0]
    for num in nums[1:]:
        curr = max(num, curr + num)
        max_sum = max(max_sum, curr)
    return max_sum

print(max_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))  # 6


# # Rotacionar array
def rotate_array(arr, k):
    k %= len(arr)
    return arr[-k:] + arr[:-k]

print(rotate_array([1,2,3,4,5,6,7], 3))  # [5,6,7,1,2,3,4]


# # Listas Ligadas:
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# # Reverter lista ligada
def reverse_list(head):
    prev = None
    while head:
        nxt = head.next
        head.next = prev
        prev = head
        head = nxt
    return prev


# # Detectar ciclo (Floyd’s Cycle)
def has_cycle(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False


# # Mesclar duas listas ordenadas
def merge_lists(l1, l2):
    dummy = tail = ListNode()
    while l1 and l2:
        if l1.val < l2.val:
            tail.next, l1 = l1, l1.next
        else:
            tail.next, l2 = l2, l2.next
        tail = tail.next
    tail.next = l1 or l2
    return dummy.next


# Pilhas e Filas:

# # Implementar pilha/filas usando arrays ou listas
stack = []
stack.append(1)
stack.append(2)
print(stack.pop())  # 2


# # Pilha com valor mínimo em O(1)
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, x):
        self.stack.append(x)
        if not self.min_stack or x <= self.min_stack[-1]:
            self.min_stack.append(x)

    def pop(self):
        if self.stack.pop() == self.min_stack[-1]:
            self.min_stack.pop()

    def get_min(self):
        return self.min_stack[-1]


# # Fila circular
class CircularQueue:
    def __init__(self, k):
        self.q = [None]*k
        self.size = k
        self.head = self.tail = -1

    def enQueue(self, val):
        if (self.tail + 1) % self.size == self.head:
            return False
        if self.head == -1: self.head = 0
        self.tail = (self.tail + 1) % self.size
        self.q[self.tail] = val
        return True

File: algorithms/_test/estrutura-de-dados-arrays-strings.rb
# Arrays e Strings em Ruby

# Inverter string
s = "hello"
puts s.reverse  # => "olleh"

# Encontrar subarray com soma máxima (Kadane)
def max_subarray(nums)
  max_sum = curr = nums.first
  nums.drop(1).each do |num|
    curr    = [num, curr + num].max
    max_sum = [max_sum, curr].max
  end
  max_sum
end

puts max_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4])  # => 6

# Rotacionar array
def rotate_array(arr, k)
  k %= arr.size
  arr[-k, k] + arr[0, arr.size - k]
end

p rotate_array([1,2,3,4,5,6,7], 3)  # => [5,6,7,1,2,3,4]


# Listas Ligadas em Ruby
class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, nxt = nil)
    @val  = val
    @next = nxt
  end
end

# Reverter lista ligada
def reverse_list(head)
  prev = nil
  curr = head
  while curr
    nxt = curr.next
    curr.next = prev
    prev = curr
    curr = nxt
  end
  prev
end

# Detectar ciclo (Floyd’s Cycle Detection)
def has_cycle(head)
  slow = fast = head
  while fast && fast.next
    slow = slow.next
    fast = fast.next.next
    return true if slow == fast
  end
  false
end

# Mesclar duas listas ordenadas
def merge_lists(l1, l2)
  dummy = ListNode.new
  tail  = dummy
  while l1 && l2
    if l1.val < l2.val
      tail.next, l1 = l1, l1.next
    else
      tail.next, l2 = l2, l2.next
    end
    tail = tail.next
  end
  tail.next = l1 || l2
  dummy.next
end


# Pilhas e Filas

# Pilha simples usando Array
stack = []
stack.push(1)
stack.push(2)
puts stack.pop  # => 2

# Pilha com valor mínimo em O(1)
class MinStack
  def initialize
    @stack     = []
    @min_stack = []
  end

  def push(x)
    @stack << x
    @min_stack << x if @min_stack.empty? || x <= @min_stack.last
  end

  def pop
    val = @stack.pop
    @min_stack.pop if val == @min_stack.last
    val
  end

  def top
    @stack.last
  end

  def get_min
    @min_stack.last
  end
end

# Exemplo MinStack
ms = MinStack.new
ms.push(3)
ms.push(5)
ms.push(2)
puts ms.get_min  # => 2
ms.pop
puts ms.get_min  # => 3

# Fila circular
class CircularQueue
  def initialize(k)
    @queue = Array.new(k)
    @size  = k
    @head  = -1
    @tail  = -1
  end

  def en_queue(val)
    return false if (@tail + 1) % @size == @head
    @head = 0 if @head == -1
    @tail = (@tail + 1) % @size
    @queue[@tail] = val
    true
  end

  def de_queue
    return false if empty?
    if @head == @tail
      @head = @tail = -1
    else
      @head = (@head + 1) % @size
    end
    true
  end

  def front
    empty? ? nil : @queue[@head]
  end

  def rear
    empty? ? nil : @queue[@tail]
  end

  def empty?
    @head == -1
  end

  def full?
    (@tail + 1) % @size == @head
  end
end

# Exemplo CircularQueue
cq = CircularQueue.new(3)
puts cq.en_queue(1)  # => true
puts cq.en_queue(2)  # => true
puts cq.en_queue(3)  # => true
puts cq.en_queue(4)  # => false (full)
puts cq.rear         # => 3
puts cq.de_queue     # => true
puts cq.en_queue(4)  # => true
puts cq.rear         # => 4
